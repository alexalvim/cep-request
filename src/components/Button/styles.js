import styled from 'styled-components';

import Colors from '../../colors';
import Spaces from '../../spaces';
import Fonts from '../../fonts';

const Container = styled.button`
  background-color: ${Colors.darkGray};
  border: none;
  color: ${Colors.darkestColor};
  cursor: pointer;
  display: inline-block;
  padding: ${Spaces.base} ${Spaces.double};
`;

const Label = styled.span`
  color: ${Colors.lightestColor};
  font-size: ${Fonts.medium};
`;

export {
  Container,
  Label
};