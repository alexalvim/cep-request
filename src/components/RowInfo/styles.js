import styled from 'styled-components';

import Colors from '../../colors';
import Spaces from '../../spaces';
import Fonts from '../../fonts';

const Row = styled.div`
  margin: ${Spaces.base} 0;
`;

const Value = styled.span`
  color: ${Colors.darkGray};
  font-size: ${Fonts.medium};
`;

const Label = styled(Value)`
  font-weight: bold;
  margin-right: ${Spaces.half};
`;

export {
  Row,
  Label,
  Value,
};