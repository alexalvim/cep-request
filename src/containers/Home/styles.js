import styled from 'styled-components';

import Colors from '../../colors';
import Spaces from '../../spaces';
import Fonts from '../../fonts';

const Container = styled.div`
  align-items: center;
  background-color: ${Colors.lightGray};
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: flex-start;
  padding: ${Spaces.double};
  width: 100%;
`;

const ContentBox = styled.div`
  background-color: ${Colors.lightestColor};
  box-shadow: 1px 2px 4px 1px ${Colors.mediumGray};
  margin: ${Spaces.base} 0;
  max-width: 400px;
  padding: ${Spaces.double};
  width: 100%;
`;

const CEPInput = styled.input`
  border: solid 1px ${Colors.darkGray};
  color: ${Colors.darkestColor};
  font-size: ${Fonts.medium};
  flex-grow: 1;
  margin-right: ${Spaces.base};
  padding: ${Spaces.base};
`;

const FormWrapper = styled.div`
  display: flex;
`;

const InfoText = styled.span`
  color: ${Colors.darkGray};
  display: block;
  font-size: ${Fonts.medium};
`;

export {
  Container,
  ContentBox,
  CEPInput,
  FormWrapper,
  InfoText,
};