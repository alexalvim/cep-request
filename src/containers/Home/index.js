import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { cepFetchRequest } from '../../actions/cep';

import RowInfo from '../../components/RowInfo';
import Button from '../../components/Button';
import { Container, ContentBox, CEPInput, FormWrapper, InfoText } from './styles';

export class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputCep: '',
    }
  }

  handleChangeCepInput = (evt) => {
    const cep = evt.currentTarget.value;
    if(cep.length <= 8) {
      this.setState({
        inputCep: cep
      });
    }
  }

  handleSubmitCep = (evt) => {
    evt.preventDefault();
    if(this.state.inputCep.length >= 7) {
      this.props.cepFetchRequest(this.state.inputCep);
      this.setState({
        inputCep: ''
      });
    } else {
      alert('Digite um CEP v√°lido!');
    }
  }

  render() {
    const { inputCep } = this.state;
    const { handleChangeCepInput, handleSubmitCep } = this;
    const { cepNumber, uf, locale, place, isLoading, hasError } = this.props;
    return (
      <Container>
        <ContentBox>
          <form
            onSubmit={handleSubmitCep}>
            <FormWrapper>
              <CEPInput
                type='number'
                value={inputCep}
                onChange={handleChangeCepInput} />
              <Button
                label='Buscar CEP'
                onClick={handleSubmitCep} />
            </FormWrapper>
          </form>
        </ContentBox>
        {isLoading && <InfoText>Carregando</InfoText>}
        {!isLoading && ((cepNumber && cepNumber) || hasError) &&
          <ContentBox>
            {!isLoading && cepNumber && cepNumber !== '' && 
              <Fragment>
                <RowInfo
                  label='CEP'
                  value={cepNumber} />
                {uf && uf !== '' &&
                  <RowInfo
                    label='Estado'
                    value={uf} />
                }
                {locale && locale !== '' &&
                  <RowInfo
                    label='Cidade'
                    value={locale} />
                }
                {place && place !== '' &&
                  <RowInfo
                    label='Logradouro'
                    value={place} />
                }
              </Fragment>
            }
            {!isLoading && hasError && <InfoText>Erro ao buscar CEP</InfoText>}
          </ContentBox>
        }
      </Container>
    );
  }
}

const mapStateToProps = store => ({
  cepNumber: store.cepState.cepNumber,
  uf: store.cepState.uf,
  locale: store.cepState.locale,
  place: store.cepState.place,
  hasError: store.cepState.hasError,
  isLoading: store.cepState.isLoading,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ cepFetchRequest }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Home);
