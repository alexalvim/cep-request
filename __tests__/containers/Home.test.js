import React from 'react';
import { shallow } from 'enzyme';

import { Home } from '../../src/containers/Home';
import { CEPInput, InfoText } from '../../src/containers/Home/styles';
import RowInfo from '../../src/components/RowInfo';

describe('Testing Home', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render input', () => {
    const wrapper = shallow(<Home />);

    expect(wrapper.find(CEPInput)).toHaveLength(1);
  });

  it('should render loading text', () => {
    const wrapper = shallow(<Home isLoading={true} />);

    expect(wrapper.find(InfoText).text()).toEqual('Carregando');
  });

  it('should render error text', () => {
    const wrapper = shallow(<Home isLoading={false} hasError={true} />);

    expect(wrapper.find(InfoText).text()).toEqual('Erro ao buscar CEP');
  });

  it('should render one RowInfo when there is only one information to show', () => {
    const wrapper = shallow(<Home isLoading={false} cepNumber='00000000' />);

    expect(wrapper.find(RowInfo)).toHaveLength(1);
  });

  it('should render all RowInfo when there are all of informations', () => {
    const wrapper = shallow(<Home
                              isLoading={false}
                              cepNumber='00000000'
                              uf='foo'
                              locale='bar'
                              place='baz' />);

    expect(wrapper.find(RowInfo)).toHaveLength(4);
  });

  it('should change the state value when call handleChangeCepInput', () => {
    const mockedEvt = {
      currentTarget: {
        value: 'foo',
      },
    };
    const wrapper = shallow(<Home/>);
    expect(wrapper.instance().state.inputCep).toEqual('');
    wrapper.instance().handleChangeCepInput(mockedEvt);
    expect(wrapper.instance().state.inputCep).toEqual('foo');
  });

  it('in handleSubmitCep should call cepFetchRequest when the value is valid', () => {
    const mockedFunction = jest.fn();
    const evt = {
      preventDefault: () => 'foo'
    }
    const wrapper = shallow(<Home cepFetchRequest={mockedFunction} />);
    wrapper.instance().setState({ inputCep: '00000000' });
    wrapper.instance().handleSubmitCep(evt);
    expect(mockedFunction).toHaveBeenCalled();
    expect(wrapper.instance().state.inputCep).toEqual('');
  });

  it('in handleSubmitCep should show an alert when the value is invalid', () => {
    const evt = {
      preventDefault: () => 'foo'
    }
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const wrapper = shallow(<Home/>);
    wrapper.instance().setState({ inputCep: '000' });
    wrapper.instance().handleSubmitCep(evt);
    expect(window.alert).toBeCalledWith('Digite um CEP v√°lido!');
  });
});