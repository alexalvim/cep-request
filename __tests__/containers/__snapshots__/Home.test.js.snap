// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Home should render correctly 1`] = `ShallowWrapper {}`;
